MoreLinq/IndexBy.cs

```diff
@@ -0,0 +1,73 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Applies a key-generating function to each element of a sequence and returns a sequence that
+        /// contains the elements of the original sequence as well its key and index inside the group of its key.
+        /// </summary>
+        /// <typeparam name="TSource">Type of the elements of the source sequence.</typeparam>
+        /// <typeparam name="TKey">Type of the projected key.</typeparam>
+        /// <typeparam name="TResult">Type of the projected element.</typeparam>
+        /// <param name="source">Source sequence.</param>
+        /// <param name="keySelector">Function that transforms each item of source sequence into a key to be compared against the others.</param>
+        /// <param name="resultSelector">The projection to each element, its key and index.</param>
+        /// <returns>A sequence of unique keys and their number of occurrences in the original sequence.</returns>
+
+        public static IEnumerable<TResult> IndexBy<TSource, TKey, TResult>(
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            Func<TSource, TKey, int, TResult> resultSelector)
+        {
+            return IndexBy(source, keySelector, resultSelector, null);
+        }
+
+        /// <summary>
+        /// Applies a key-generating function to each element of a sequence and returns a sequence that
+        /// contains the elements of the original sequence as well its key and index inside the group of its key.
+        /// An additional argument specifies a comparer to use for testing equivalence of keys.
+        /// </summary>
+        /// <typeparam name="TSource">Type of the elements of the source sequence.</typeparam>
+        /// <typeparam name="TKey">Type of the projected key.</typeparam>
+        /// <typeparam name="TResult">Type of the projected element.</typeparam>
+        /// <param name="source">Source sequence.</param>
+        /// <param name="keySelector">Function that transforms each item of source sequence into a key to be compared against the others.</param>
+        /// <param name="resultSelector">The projection to each element, its key and index.</param>
+        /// <param name="comparer">The equality comparer to use to determine whether or not keys are equal.
+        /// If null, the default equality comparer for <typeparamref name="TSource"/> is used.</param>
+        /// <returns>A sequence of unique keys and their number of occurrences in the original sequence.</returns>
+
+        public static IEnumerable<TResult> IndexBy<TSource, TKey, TResult>(
```

Comment:
I feel like `IndexBy` is not very descriptive of what this does. Other languages like Ruby have `index_by` as what C# knows as `ToDictionary`. I'm having a hard time coming up with a better name. Since this is mostly `CountBy`, where the last parameter to `resultSelector` is decremented once, does this really make sense to add?

Reply:
> I feel like `IndexBy` is not very descriptive of what this does. Other languages like Ruby have `index_by` as what C# knows as `ToDictionary`. I'm having a hard time coming up with a better name.

It's consistent with [`Index`](https://morelinq.github.io/3.1/ref/api/html/Overload_MoreLinq_MoreEnumerable_Index.htm), which pairs each element of sequence with its position in the sequence. `IndexBy` does the same except the positions are relative to a key group. That's how the name makes sense to me; that it's a variation of `Index`.

> Since this is mostly `CountBy`, where the last parameter to `resultSelector` is decremented once, does this really make sense to add?

`CountBy` aggregates on keys. This gives you the source elements in the result so is different enough to be useful.